File: app/services/analysis_storage_service.py
"""
Analysis Storage Service for TrustLens (Imagine Cup)

This service stores analysis results in Azure Cosmos DB using cryptographic hashes.

    Privacy & Responsible AI Design:
    - Only hash values are stored as document IDs; no raw images or text are persisted.
    - This approach protects user privacy by ensuring that raw sensitive content never leaves the ephemeral memory of the server for long-term storage.
    - By using cryptographic hashes (SHA-256), we can detect duplicates without knowing the original content.
    - We do not infer location, authorship, or personally identifiable information, adhering to Imagine Cup ethical standards.

"""

from datetime import datetime, timezone
from azure.cosmos import CosmosClient, PartitionKey, exceptions
from app.config.settings import Config


_cosmos_client = None
_container = None


def _get_container():
    """
    Lazily initialize and return the Cosmos DB container.
    Uses hash as the partition key for efficient lookups.
    """
    global _cosmos_client, _container
    
    if _container is not None:
        return _container
    
    if not Config.COSMOS_ENDPOINT or not Config.COSMOS_KEY:
        print("⚠️ Cosmos DB credentials not configured. Storage disabled.")
        return None
    
    try:
        _cosmos_client = CosmosClient(Config.COSMOS_ENDPOINT, Config.COSMOS_KEY)
        
        database = _cosmos_client.create_database_if_not_exists(id=Config.COSMOS_DATABASE)
        
        _container = database.create_container_if_not_exists(
            id=Config.COSMOS_CONTAINER,
            partition_key=PartitionKey(path="/hash")
        )
        
        print(f"✅ Connected to Cosmos DB: {Config.COSMOS_DATABASE}/{Config.COSMOS_CONTAINER}")
        return _container
    except Exception as e:
        print(f"❌ Cosmos DB connection failed: {str(e)}")
        return None


def store_analysis(hash_value: str, data_type: str, analysis_result: dict) -> dict:
    """
    Store an analysis result in Cosmos DB.
    
    Args:
        hash_value: SHA-256 hash of the content (used as document id and partition key)
        data_type: "image" or "text"
        analysis_result: The analysis result object (no raw content)
    
    Returns:
        The stored document or error info
    
    Privacy Note: Only the hash is stored as the identifier.
    Raw user content is never persisted to protect privacy.
    """
    container = _get_container()
    if container is None:
        return {"success": False, "error": "Cosmos DB not configured"}
    
    document = {
        "id": hash_value,
        "hash": hash_value,
        "type": data_type,
        "analysis": analysis_result,
        "createdAt": datetime.now(timezone.utc).isoformat()
    }
    
    try:
        container.upsert_item(document)
        print(f"✅ Stored analysis for hash: {hash_value[:16]}...")
        return {"success": True, "document": document}
    except exceptions.CosmosHttpResponseError as e:
        print(f"❌ Failed to store analysis: {str(e)}")
        return {"success": False, "error": str(e)}


def get_analysis_by_hash(hash_value: str) -> dict:
    """
    Retrieve a previously stored analysis by its hash.
    
    Args:
        hash_value: SHA-256 hash of the content
    
    Returns:
        The stored document if found, None otherwise
    
    This enables efficient deduplication: if we've seen this content before,
    we return the cached analysis instead of re-processing.
    """
    container = _get_container()
    if container is None:
        return None
    
    try:
        item = container.read_item(item=hash_value, partition_key=hash_value)
        print(f"✅ Found existing analysis for hash: {hash_value[:16]}...")
        return item
    except exceptions.CosmosResourceNotFoundError:
        return None
    except exceptions.CosmosHttpResponseError as e:
        print(f"⚠️ Error retrieving analysis: {str(e)}")
        return None

File: src/services/analysisStorage.service.js
import { CosmosClient } from "@azure/cosmos";
import dotenv from "dotenv";

dotenv.config();

const endpoint = process.env.COSMOS_ENDPOINT;
const key = process.env.COSMOS_KEY;
const databaseId = "trustlensDB";
const containerId = "analysis_records";

// Initialize Cosmos Client only if credentials are provided
let container = null;

if (endpoint && key) {
  try {
    const client = new CosmosClient({ endpoint, key });
    const database = client.database(databaseId);
    container = database.container(containerId);
    console.log("✅ Cosmos DB client initialized");
  } catch (error) {
    console.error("❌ Failed to initialize Cosmos DB client:", error.message);
  }
} else {
  console.warn("⚠️ Cosmos DB credentials missing. Storage service will operate in mock mode.");
}

/**
 * Stores an analysis result in Cosmos DB.
 * Uses hashes to identify content without storing raw user data.
 * This protects user privacy by ensuring that sensitive images or text
 * are never persisted in their original form.
 * 
 * @param {Object} params - Storage parameters
 * @param {string} params.hash - Unique hash of the content
 * @param {string} params.type - Content type ("image" or "text")
 * @param {Object} params.analysis - The analysis result to store
 * @returns {Promise<Object>} The stored document
 */
export async function storeAnalysis({ hash, type, analysis }) {
  if (!container) {
    console.warn("Storage skipped: Cosmos DB container not initialized");
    return null;
  }

  const document = {
    id: hash, // Use hash as ID for direct lookup
    hash: hash,
    type: type,
    analysis: analysis,
    createdAt: new Date().toISOString(),
    responsibleAI: "TrustLens Privacy Shield - No raw content stored"
  };

  try {
    const { resource } = await container.items.create(document);
    return resource;
  } catch (error) {
    // If it already exists (Conflict 409), we just return null or handle accordingly
    if (error.code === 409) {
      console.log(`Document with hash ${hash} already exists.`);
      return null;
    }
    console.error("Error storing analysis:", error.message);
    throw error;
  }
}

/**
 * Finds an analysis result by its content hash.
 * This allows us to reuse results for identical content (Image Tracing).
 * 
 * @param {string} hash - The hash to search for
 * @returns {Promise<Object|null>} The stored analysis or null if not found
 */
export async function findAnalysisByHash(hash) {
  if (!container) return null;

  try {
    const { resource } = await container.item(hash, hash).read();
    return resource || null;
  } catch (error) {
    // If not found (404), return null
    if (error.code === 404) return null;
    console.error("Error finding analysis by hash:", error.message);
    return null;
  }
}
